enhancements/ingress/request-timeout-configuration.md

```diff
@@ -0,0 +1,231 @@
+shortening the timeout, allowing connections to be closed more aggressively.
+
+### Goals
+
+Allow admins to configure the following HAProxy connection timeout parameters:
+- timeout client
+- timeout client-fin
+- timeout server
+- timeout server-fin
+- tcp-request inspect-delay
+
+### Non-Goals
+
+- Allow admins to configure every HAProxy timeout parameter
+
+## Proposal
+
+Add the following configurable variables to the ingresscontroller API:
+```yaml
+spec:
+  tuningOptions:
+    clientTimeout: "30s"
+    clientFinTimeout: "1s"
+    serverTimeout: "30s"
+    serverFinTimeout: "1s"
+    tlsInspectDelay: "10s"
+```
+
+All variables expect time values. If no units are specified, the router assumes
+milliseconds.
+
+### User Stories
+
+TODO
+
+### Implementation Details/Notes/Constraints
+
+#### Router Image Environment Variables
+
+`timeout client`, `timeout client-fin`, `timeout server`, and `timeout
+server-fin` were all configurable in 3.X, and the router image still supports
+configuring them from the environment variables
+`ROUTER_DEFAULT_CLIENT_TIMEOUT`, `ROUTER_CLIENT_FIN_TIMEOUT`,
+`ROUTER_DEFAULT_SERVER_TIMEOUT`, and `ROUTER_DEFAULT_SERVER_FIN_TIMEOUT`,
+respectively.
+
+A new environment variable will be defined, `ROUTER_INSPECT_DELAY`, which will
+control the `tcp-request inspect-delay` variable.
+
+All the timeout environment variables expect values to follow the HAProxy time
+format. The format is an integer optionally followed by a time unit. If no unit
+is specified, the value represents the delay in milliseconds. The full
+description of the time format can be found in [section 2.4. of the HAProxy 2.2
+documentation](https://github.com/haproxy/haproxy/blob/v2.2.0/doc/configuration.txt).
+
+#### Time Format Validation
+
+The format specified in [section 2.4. of the HAProxy 2.2
+documentation](https://github.com/haproxy/haproxy/blob/v2.2.0/doc/configuration.txt)
+will be enforced when the ingresscontroller config is admitted using the
+regular expression `[0-9]+(?:[um]?s|[mhd])?`. The regular expression should
+accept strings containing one or more digits, optionally followed by one of the
+following time units:
+- us
+- ms
+- s
+- m
+- h
+- d
+
+In addition, HAProxy only allows a maximum timeout of 2,147,483,647ms (about 24
+days, 20 hours). If a time greater than the maximum is specified for any of the
+API fields added in this proposal, the timeout will be set to the maximum
+allowed value, and an error message will be logged by the ingress operator.
+
+### Risks and Mitigations
+
+Setting timeout values in the ingress controller will only affect timeouts
+caused by ingress. Other hops in the connection, such as the backend or the
+cloud LB, may still cause connections to time out even if the fields added in
+this proposal are set sufficiently high.
+
+## Design Details
+
+### Open Questions
+
+- How do these timeout variables affect websocket connections (if at all)?
```

Comment:
How about `hard-stop-after`?  We added an annotation to configure `hard-stop-after` in https://github.com/openshift/cluster-ingress-operator/pull/522, but a formal API field would be useful.

Reply:
It would probably be good to implement `hard-stop-after` in the API, but since it's already configurable with the annotation, let's leave that addition for a later date.

Reply:
Might be worth adding as a non-goal.